api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,307 @@
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to an HttpHeaders message.
+message HeadersResponse {
+  CommonResponse response = 1;
+}
+
+// This message must be sent in response to an HttpTrailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+// This message must be sent in response to an HttpBody message.
+message BodyResponse {
+  CommonResponse response = 1;
+}
+
+// This message contains common fields between header and body responses.
+// [#next-free-field: 6]
+message CommonResponse {
+  enum ResponseStatus {
+    // Apply the mutation instructions in this message to the
+    // request or response, and then continue processing the filter
+    // stream as normal. This is the default.
+    CONTINUE = 0;
+
+    // Replace the request or response with the contents
+    // of this message. If header_mutation is set, apply it to the
+    // headers. If body_mutation is set and contains a body, then add that
+    // body to the request or response, even if one does not already exist --
+    // otherwise, clear the body. Any additional body and trailers
+    // received from downstream or upstream will be ignored.
+    // This can be used to add a body to a request or response that does not
+    // have one already.
+    CONTINUE_AND_DO_NOT_END_STREAM = 1;
+  }
+
+  // If set, provide additional direction on how the Envoy proxy should
+  // handle the rest of the HTTP filter chain.
+  ResponseStatus status = 1 [(validate.rules).enum = {defined_only: true}];
+
+  // Instructions on how to manipulate the headers. When responding to an
+  // HttpBody request, header mutations will only take effect if the
+  // headers were not already sent further on the filter chain, which
+  // happens only if the current processing mode for the body is BUFFERED
+  // or BUFFERED_PARTIAL.
+  HeaderMutation header_mutation = 2;
+
+  // Replace the body of the last message sent to the remote server on this
+  // stream. If responding to an HttpBody request, simply replace or clear
+  // the body chunk that was sent with that request. If responding to an
+  // HttpHeaders request, then a new body may be added to the request if this
+  // message is returned along with the CONTINUE_AND_DO_NOT_END_STREAM status.
+  BodyMutation body_mutation = 3;
+
+  // Add new trailers to the message. This may be used when responding to either a
+  // HttpHeaders or HttpBody message, but only if this message is returned
+  // along with the CONTINUE_AND_DO_NOT_END_STREAM status.
+  config.core.v3.HeaderMap trailers = 4;
+
+  // Clear the route cache for the current request.
```

Comment:
From a wire perspective, I think this is leaking Envoy implementation details out. Should this be something more imperative, e.g. `recalculate_route`? CC @markdroth in case gRPC ever implements something like this protocol.

Reply:
That's the same name that we use in the ext_proc filter, but another name is fine with me.